src/mongo/db/auth/authorization_session.h

```diff
@@ -145,7 +145,7 @@ class AuthorizationSession {
      * Adds the User identified by "UserName" to the authorization session, acquiring privileges
      * for it in the process.
      */
-    virtual Status addAndAuthorizeUser(OperationContext* opCtx, const UserName& userName) = 0;
+    virtual Status addAndAuthorizeUser(OperationContext* opCtx, const UserName& userName, bool fromSecurityToken = false) = 0;
```

Comment:
Actually, just thought about this some more in context of my concerns about the signature validation not being performed reliably.

How about if we have two interface methods:

```
/**
 * Connection based authentication using `saslStart/saslContinue` or `authenticate`
 */
 virtual Status addAndAuthorizeUser(OperationContext*, const UserName&) = 0;
 
 /**
 * Security Token based authentication used with mutltitenancy support.
 */
virtual Status addAndAuthorizeUser(OperationContext*, const SecurityToken&) = 0;
```

Then the signature validation happens at the same time as authorization grants, which will dramatically reduce the chances that we fail at this.

Reply:
I'm tying to synthesize your comments in order to implement the changes, but I don't think I'm following exactly what you're suggesting - my current understanding is that we could make changes that look something like:
-  Change `verifySecurityToken(tokenObj)` to `parseSecurityToken(tokenObj)`. This calls SecurityToken::parse(), then validates the signature.
- Create some type of wrapper for SecurityToken and put that on the request object that looks something like:
```
class SecurityTokenWrapper {
public:
    SecurityTokenWrapper(const BSONObj& tokenObj) : _tokenObj(tokenObj) {}
 
    TenantId getTenantId() {
        if (!token) {
            _token = parseSecurityToken(_tokenObj);
        }
       return _token->getAuthenticatedUser().getTenant();
    }
   
    getAuthenticatedUser() { ... looks similar to getTenantId}
private:
    BSONObj _tokenObj;
    SecurityToken* _token;
}
```
- In command request parsers, the call to parseNSCollectionRequired would now be `ctxt.parseNSCollectionRequired(_dbName, commandElement, _securityTokenWrapper.getTenantId())`
- Implement the changes you mention in this comment to create two `addAndAuthorizeUser` methods.

The main thing I'm confused about is where we'd want to construct the `SecurityTokenAuthenticationGuard` and call `addAndAuthorizeUser` - for signature validation to happen at the same time as authz grants as you mentioned, we'd need to call `addAndAuthorizeUser` when we call` parseSecurityToken`. We'll already have constructed and parsed the actual SecurityToken object by the time we construct the SecurityTokenAuthenticationGuard and do authz grants though, because we access the tenantId when parsing the command, and I don't think we can move handling the scope guard/authz grants into command parsing.